

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Peak alignment by dynamic programming &mdash; PyMassSpec 2.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> PyMassSpec
          

          
            
            <img src="../_static/pyms.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                2.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="chapter01.html">1. Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../pyms/documentation.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Baseline/Baseline.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Baseline</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Deconvolution/Deconvolution.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Deconvolution</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Display/Display.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Display</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Experiment/Experiment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Experiment</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Gapfill/Gapfill.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Gapfill</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/GCMS/GCMS.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.GCMS</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Noise/Noise.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Noise</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Peak/Peak.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Peak</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Simulator/Simulator.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Simulator</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms/Utils/Utils.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyms.Utils</span></code></a></li>
</ul>
<p class="caption"><span class="caption-text">Contributing to PyMS</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Contributing/Contributing.html">1. Contributing to PyMassSpec</a></li>
<li class="toctree-l1"><a class="reference internal" href="Contributing/StyleGuide.html">2. PyMassSpec coding Style Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="Contributing/Source.html">3. Downloading PyMS source code</a></li>
<li class="toctree-l1"><a class="reference internal" href="Contributing/Building.html">4. Building from Source</a></li>
</ul>
<p class="caption"><span class="caption-text">Tests and Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/20a/20a.html">20a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/20b/20b.html">20b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/20c/20c.html">20c</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/20d/20d.html">20d</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/30a/30a.html">30a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/30b/30b.html">30a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/30c/30c.html">30c</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/31/31.html">31</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/32/32.html">32</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/40a/40a.html">40a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/40b/40b.html">40b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/41a/41a.html">41a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/41b/41b.html">41b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/42a/42a.html">42a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/42b/42b.html">42b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/43/43.html">43</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/50/50.html">50</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/51/51.html">51</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/52/52.html">52</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/53/53.html">53</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/54/54.html">54</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/60/60.html">60</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/61a/61a.html">61a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/61b/61b.html">61b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/62/62.html">62</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/63/63.html">63</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/64/64.html">64</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/70a/70a.html">70a</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/70b/70b.html">70b</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/71/71.html">71</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/90/90.html">90</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/91/91.html">91</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/92/92.html">92</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/93/93.html">93</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/94/94.html">94</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/95/95.html">95</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/A1/A1.html">A1</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/A2/A2.html">A2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pyms-test/x10/x10.html">x10</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PyMassSpec</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Peak alignment by dynamic programming</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/domdfcoding/pyms/docs/UserGuide/chapter07.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="peak-alignment-by-dynamic-programming">
<h1><a class="toc-backref" href="#id5">Peak alignment by dynamic programming</a><a class="headerlink" href="#peak-alignment-by-dynamic-programming" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="table-of-contents">
<p class="topic-title first">Table of Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#peak-alignment-by-dynamic-programming" id="id5">Peak alignment by dynamic programming</a></p>
<ul>
<li><p><a class="reference internal" href="#preparation-of-multiple-experiments-for-peak-alignment-by-dynamic-programming" id="id6">Preparation of multiple experiments for peak alignment by dynamic programming</a></p>
<ul>
<li><p><a class="reference internal" href="#creating-an-experiment" id="id7">Creating an Experiment</a></p></li>
<li><p><a class="reference internal" href="#multiple-experiments" id="id8">Multiple Experiments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#dynamic-programming-alignment-of-peak-lists-from-multiple-experiments" id="id9">Dynamic programming alignment of peak lists from multiple experiments</a></p></li>
<li><p><a class="reference internal" href="#between-state-alignment-of-peak-lists-from-multiple-experiments" id="id10">Between-state alignment of peak lists from multiple experiments</a></p></li>
<li><p><a class="reference internal" href="#common-ion-area-quantitation" id="id11">Common Ion Area Quantitation</a></p>
<ul>
<li><p><a class="reference internal" href="#using-the-common-ion-algorithm" id="id12">Using the Common Ion Algorithm</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#references" id="id13">References</a></p></li>
</ul>
</li>
</ul>
</div>
<p>PyMS provides functions to align GC-MS peaks by dynamic programming
<a class="footnote-reference brackets" href="#id4" id="id1">1</a>.  The peak alignment by dynamic programming uses both peak
apex retention time and mass spectra.  This information is determined from the
raw GC-MS data by applying a series of processing steps to produce data that
can then be aligned and used for statistical analysis.  The details are
described in this chapter.</p>
<div class="section" id="preparation-of-multiple-experiments-for-peak-alignment-by-dynamic-programming">
<h2><a class="toc-backref" href="#id6">Preparation of multiple experiments for peak alignment by dynamic programming</a><a class="headerlink" href="#preparation-of-multiple-experiments-for-peak-alignment-by-dynamic-programming" title="Permalink to this headline">¶</a></h2>
<div class="section" id="creating-an-experiment">
<h3><a class="toc-backref" href="#id7">Creating an Experiment</a><a class="headerlink" href="#creating-an-experiment" title="Permalink to this headline">¶</a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This example is in <a class="reference external" href="../pyms-test/60/60.html">pyms-test/60</a></p>
</div>
<p>Before aligning peaks from multiple experiments, the peak objects need to be
created and encapsulated into PyMS experiment objects. During this process
it is often useful to pre-process the peaks in some way, for example to
null certain m/z channels and/or to select a certain retention time range.</p>
<p>To capture the data and related information prior to peak alignment, an
<a class="reference internal" href="../pyms/Experiment/Experiment.html#pyms.Experiment.Class.Experiment" title="pyms.Experiment.Class.Experiment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Experiment</span></code></a>
object is used. The <a class="reference internal" href="../pyms/Experiment/Experiment.html#pyms.Experiment.Class.Experiment" title="pyms.Experiment.Class.Experiment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Experiment</span></code></a>
object is defined in <a class="reference internal" href="../pyms/Experiment/Experiment.html#module-pyms.Experiment.Class" title="pyms.Experiment.Class"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyms.Experiment.Class</span></code></a>.</p>
<p>The procedure is to proceed as described in the previous chapter. Namely: read
a file; bin the data into fixed mass values; smooth the data; remove the
baseline; deconvolute peaks; filter the peaks; set the mass range; remove
uninformative ions; and estimate peak areas. The process is given in the
following program listing.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54</pre></div></td><td class="code"><div class="highlight"><pre><span></span> <span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>

 <span class="kn">from</span> <span class="nn">pyms.GCMS.IO.ANDI.Function</span> <span class="kn">import</span> <span class="n">ANDI_reader</span>
 <span class="kn">from</span> <span class="nn">pyms.GCMS.Function</span> <span class="kn">import</span> <span class="n">build_intensity_matrix_i</span>
 <span class="kn">from</span> <span class="nn">pyms.Noise.SavitzkyGolay</span> <span class="kn">import</span> <span class="n">savitzky_golay</span>
 <span class="kn">from</span> <span class="nn">pyms.Baseline.TopHat</span> <span class="kn">import</span> <span class="n">tophat</span>
 <span class="kn">from</span> <span class="nn">pyms.Peak.Class</span> <span class="kn">import</span> <span class="n">Peak</span>
 <span class="kn">from</span> <span class="nn">pyms.Peak.Function</span> <span class="kn">import</span> <span class="n">peak_sum_area</span>

 <span class="kn">from</span> <span class="nn">pyms.Deconvolution.BillerBiemann.Function</span> <span class="kn">import</span> <span class="n">BillerBiemann</span><span class="p">,</span> \
     <span class="n">rel_threshold</span><span class="p">,</span> <span class="n">num_ions_threshol</span>

 <span class="c1"># deconvolution and peak list filtering parameters</span>
 <span class="n">points</span> <span class="o">=</span> <span class="mi">9</span><span class="p">;</span> <span class="n">scans</span> <span class="o">=</span> <span class="mi">2</span><span class="p">;</span> <span class="n">n</span> <span class="o">=</span> <span class="mi">3</span><span class="p">;</span> <span class="n">t</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">;</span> <span class="n">r</span> <span class="o">=</span> <span class="mi">2</span><span class="p">;</span>

 <span class="n">andi_file</span> <span class="o">=</span> <span class="s2">&quot;data/a0806_077.cdf&quot;</span>

 <span class="n">data</span> <span class="o">=</span> <span class="n">ANDI_reader</span><span class="p">(</span><span class="n">andi_file</span><span class="p">)</span>

 <span class="c1"># integer mass</span>
 <span class="n">im</span> <span class="o">=</span> <span class="n">build_intensity_matrix_i</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

 <span class="c1"># get the size of the intensity matrix</span>
 <span class="n">n_scan</span><span class="p">,</span> <span class="n">n_mz</span> <span class="o">=</span> <span class="n">im</span><span class="o">.</span><span class="n">get_size</span><span class="p">()</span>

 <span class="c1"># smooth data</span>
 <span class="k">for</span> <span class="n">ii</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_mz</span><span class="p">):</span>
     <span class="n">ic</span> <span class="o">=</span> <span class="n">im</span><span class="o">.</span><span class="n">get_ic_at_index</span><span class="p">(</span><span class="n">ii</span><span class="p">)</span>
     <span class="n">ic1</span> <span class="o">=</span> <span class="n">savitzky_golay</span><span class="p">(</span><span class="n">ic</span><span class="p">)</span>
     <span class="n">ic_smooth</span> <span class="o">=</span> <span class="n">savitzky_golay</span><span class="p">(</span><span class="n">ic1</span><span class="p">)</span>
     <span class="n">ic_base</span> <span class="o">=</span> <span class="n">tophat</span><span class="p">(</span><span class="n">ic_smooth</span><span class="p">,</span> <span class="n">struct</span><span class="o">=</span><span class="s2">&quot;1.5m&quot;</span><span class="p">)</span>
     <span class="n">im</span><span class="o">.</span><span class="n">set_ic_at_index</span><span class="p">(</span><span class="n">ii</span><span class="p">,</span> <span class="n">ic_base</span><span class="p">)</span>

 <span class="c1"># do peak detection on pre-trimmed data</span>

 <span class="c1"># get the list of Peak objects</span>
 <span class="n">pl</span> <span class="o">=</span> <span class="n">BillerBiemann</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">points</span><span class="p">,</span> <span class="n">scans</span><span class="p">)</span>

 <span class="c1"># trim by relative intensity</span>
 <span class="n">apl</span> <span class="o">=</span> <span class="n">rel_threshold</span><span class="p">(</span><span class="n">pl</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>

 <span class="c1"># trim by threshold</span>
 <span class="n">peak_list</span> <span class="o">=</span> <span class="n">num_ions_threshold</span><span class="p">(</span><span class="n">apl</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>

 <span class="k">print</span> <span class="s2">&quot;Number of Peaks found:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">peak_list</span><span class="p">)</span>

 <span class="c1"># ignore TMS ions and set mass range</span>
 <span class="k">for</span> <span class="n">peak</span> <span class="ow">in</span> <span class="n">peak_list</span><span class="p">:</span>
     <span class="n">peak</span><span class="o">.</span><span class="n">crop_mass</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span><span class="mi">540</span><span class="p">)</span>
     <span class="n">peak</span><span class="o">.</span><span class="n">null_mass</span><span class="p">(</span><span class="mi">73</span><span class="p">)</span>
     <span class="n">peak</span><span class="o">.</span><span class="n">null_mass</span><span class="p">(</span><span class="mi">147</span><span class="p">)</span>
     <span class="c1"># find area</span>
     <span class="n">area</span> <span class="o">=</span> <span class="n">peak_sum_area</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">peak</span><span class="p">)</span>
     <span class="n">peak</span><span class="o">.</span><span class="n">set_area</span><span class="p">(</span><span class="n">area</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>The resulting list of peaks can now be stored as an
<a class="reference internal" href="../pyms/Experiment/Experiment.html#pyms.Experiment.Class.Experiment" title="pyms.Experiment.Class.Experiment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Experiment</span></code></a> object.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>57
58
59
60
61
62
63
64
65
66</pre></div></td><td class="code"><div class="highlight"><pre><span></span> <span class="kn">from</span> <span class="nn">pyms.Experiment.Class</span> <span class="kn">import</span> <span class="n">Experiment</span>
 <span class="kn">from</span> <span class="nn">pyms.Experiment.IO</span> <span class="kn">import</span> <span class="n">store_expr</span>

 <span class="c1"># create an experiment</span>
 <span class="n">expr</span> <span class="o">=</span> <span class="n">Experiment</span><span class="p">(</span><span class="s2">&quot;a0806_077&quot;</span><span class="p">,</span> <span class="n">peak_list</span><span class="p">)</span>

 <span class="c1"># set time range for all experiments</span>
 <span class="n">expr</span><span class="o">.</span><span class="n">sele_rt_range</span><span class="p">([</span><span class="s2">&quot;6.5m&quot;</span><span class="p">,</span> <span class="s2">&quot;21m&quot;</span><span class="p">])</span>

 <span class="n">store_expr</span><span class="p">(</span><span class="s2">&quot;output/a0806_077.expr&quot;</span><span class="p">,</span> <span class="n">expr</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>Once an experiment has been defined, it is possible to limit the peak list to
a desired range using
<a class="reference internal" href="../pyms/Experiment/Experiment.html#pyms.Experiment.Class.Experiment.sele_rt_range" title="pyms.Experiment.Class.Experiment.sele_rt_range"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sele_rt_range()</span></code></a>.
The resulting experiment object can then be stored for later alignment.</p>
</div>
<div class="section" id="multiple-experiments">
<h3><a class="toc-backref" href="#id8">Multiple Experiments</a><a class="headerlink" href="#multiple-experiments" title="Permalink to this headline">¶</a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This example is in <a class="reference external" href="../pyms-test/61a/61a.html">pyms-test/61a</a></p>
</div>
<p>This example considers the preparation of three GC-MS experiments for peak
alignment. The experiments are named <code class="docutils literal notranslate"><span class="pre">a0806_077</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_078</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_079</span></code>,
and represent separate GC-MS sample runs from the same biological sample.</p>
<p>The procedure is the same as above, and repeated for each experiment.  For
example:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span> <span class="c1"># define path to data files</span>
 <span class="n">base_path</span> <span class="o">=</span> <span class="s2">&quot;data/&quot;</span>

 <span class="c1"># define experiments to process</span>
 <span class="n">expr_codes</span> <span class="o">=</span> <span class="p">[</span> <span class="s2">&quot;a0806_077&quot;</span><span class="p">,</span> <span class="s2">&quot;a0806_078&quot;</span><span class="p">,</span> <span class="s2">&quot;a0806_079&quot;</span> <span class="p">]</span>

 <span class="c1"># loop over all experiments</span>
 <span class="k">for</span> <span class="n">expr_code</span> <span class="ow">in</span> <span class="n">expr_codes</span><span class="p">:</span>

     <span class="k">print</span> <span class="s2">&quot;Processing&quot;</span><span class="p">,</span> <span class="n">expr_code</span>

     <span class="c1"># define the names of the peak file and the corresponding ANDI-MS file</span>
     <span class="n">andi_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">base_path</span><span class="p">,</span> <span class="n">expr_code</span> <span class="o">+</span> <span class="s2">&quot;.cdf&quot;</span><span class="p">)</span>
 <span class="o">...</span>

 <span class="o">...</span>
     <span class="c1"># create an experiment</span>
     <span class="n">expr</span> <span class="o">=</span> <span class="n">Experiment</span><span class="p">(</span><span class="n">expr_code</span><span class="p">,</span> <span class="n">peak_list</span><span class="p">)</span>

     <span class="c1"># use same time range for all experiments</span>
     <span class="n">expr</span><span class="o">.</span><span class="n">sele_rt_range</span><span class="p">([</span><span class="s2">&quot;6.5m&quot;</span><span class="p">,</span> <span class="s2">&quot;21m&quot;</span><span class="p">])</span>

     <span class="n">store_expr</span><span class="p">(</span><span class="s2">&quot;output/&quot;</span><span class="o">+</span><span class="n">expr_code</span><span class="o">+</span><span class="s2">&quot;.expr&quot;</span><span class="p">,</span> <span class="n">expr</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This example is in <a class="reference external" href="../pyms-test/61b/61b.html">pyms-test/61b</a></p>
</div>
<p>The previous set of data all belong to the same experimental condition.  That
is, they represent one group and any comparison between the data is a within
group comparison. For the original experiment, another set of GC-MS data was
collected for a different experimental condition.  This group must also be
stored as a set of experiments, and can be used for between group comparison.</p>
<p>The experiments are named <code class="docutils literal notranslate"><span class="pre">a0806_140</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_141</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_142</span></code>, and are
processed and stored as above (see pyms-test/61b).</p>
</div>
</div>
<div class="section" id="dynamic-programming-alignment-of-peak-lists-from-multiple-experiments">
<h2><a class="toc-backref" href="#id9">Dynamic programming alignment of peak lists from multiple experiments</a><a class="headerlink" href="#dynamic-programming-alignment-of-peak-lists-from-multiple-experiments" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>This example is in pyms-test/62</p></li>
<li><p>This example uses the subpackage pyms.Peak.List.DPA, which in turn uses the Python package ‘Pycluster’.  For ‘Pycluster’ installation instructions see the Section ref{subsec:pycluster}</p></li>
</ul>
<p>In this example the experiments <code class="docutils literal notranslate"><span class="pre">a0806_077</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_078</span></code>, and <code class="docutils literal notranslate"><span class="pre">a0806_079</span></code>
prepared in pyms-test/61a will be aligned, and therefore the script
pyms-test/61a/proc.py must be run first, to create the files
<code class="docutils literal notranslate"><span class="pre">a0806_077.expr</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_078.expr</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_079.expr</span></code> in the directory
pyms-test/61a/output/. These files contain the post-processed peak lists
from the three experiments.</p>
<p>A script for running the dynamic programming alignment on these experiments is
given below.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29</pre></div></td><td class="code"><div class="highlight"><pre><span></span> <span class="sd">&quot;&quot;&quot;proc.py</span>
<span class="sd"> &quot;&quot;&quot;</span>

 <span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>

 <span class="kn">from</span> <span class="nn">pyms.Experiment.IO</span> <span class="kn">import</span> <span class="n">load_expr</span>
 <span class="kn">from</span> <span class="nn">pyms.Peak.List.DPA.Class</span> <span class="kn">import</span> <span class="n">PairwiseAlignment</span>
 <span class="kn">from</span> <span class="nn">pyms.Peak.List.DPA.Function</span> <span class="kn">import</span> <span class="n">align_with_tree</span><span class="p">,</span> <span class="n">exprl2alignment</span>

 <span class="c1"># define the input experiments list</span>
 <span class="n">exprA_codes</span> <span class="o">=</span> <span class="p">[</span> <span class="s2">&quot;a0806_077&quot;</span><span class="p">,</span> <span class="s2">&quot;a0806_078&quot;</span><span class="p">,</span> <span class="s2">&quot;a0806_079&quot;</span> <span class="p">]</span>

 <span class="c1"># within replicates alignment parameters</span>
 <span class="n">Dw</span> <span class="o">=</span> <span class="mf">2.5</span>  <span class="c1"># rt modulation [s]</span>
 <span class="n">Gw</span> <span class="o">=</span> <span class="mf">0.30</span> <span class="c1"># gap penalty</span>

 <span class="c1"># do the alignment</span>
 <span class="k">print</span> <span class="s1">&#39;Aligning expt A&#39;</span>
 <span class="n">expr_list</span> <span class="o">=</span> <span class="p">[]</span>
 <span class="n">expr_dir</span> <span class="o">=</span> <span class="s2">&quot;../61a/output/&quot;</span>
 <span class="k">for</span> <span class="n">expr_code</span> <span class="ow">in</span> <span class="n">exprA_codes</span><span class="p">:</span>
     <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">expr_dir</span><span class="p">,</span> <span class="n">expr_code</span> <span class="o">+</span> <span class="s2">&quot;.expr&quot;</span><span class="p">)</span>
     <span class="n">expr</span> <span class="o">=</span> <span class="n">load_expr</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
     <span class="n">expr_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">expr</span><span class="p">)</span>
 <span class="n">F1</span> <span class="o">=</span> <span class="n">exprl2alignment</span><span class="p">(</span><span class="n">expr_list</span><span class="p">)</span>
 <span class="n">T1</span> <span class="o">=</span> <span class="n">PairwiseAlignment</span><span class="p">(</span><span class="n">F1</span><span class="p">,</span> <span class="n">Dw</span><span class="p">,</span> <span class="n">Gw</span><span class="p">)</span>
 <span class="n">A1</span> <span class="o">=</span> <span class="n">align_with_tree</span><span class="p">(</span><span class="n">T1</span><span class="p">,</span> <span class="n">min_peaks</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

 <span class="n">A1</span><span class="o">.</span><span class="n">write_csv</span><span class="p">(</span><span class="s1">&#39;output/rt.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;output/area.csv&#39;</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>The script reads the experiment files from the directory where they were stored
(<code class="docutils literal notranslate"><span class="pre">61a/output</span></code>), and creates a list of the loaded <code class="docutils literal notranslate"><span class="pre">Experiment</span></code> objects.
Each experiment object is converted into an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object with {tt
exprl2alignment()``. In this example, there is only one experimental condition
so the alignment object is only for within group alignment (this special case
is called 1-alignment). The variable F1 is a Python list containing three
alignment objects.</p>
<p>The pairwise alignment is then performed.  The parameters for the alignment by
dynamic programming are: Dw, the retention time modulation in seconds; and Gw,
the gap penalty.  These parameters are explained in detail in
<a class="footnote-reference brackets" href="#id4" id="id2">1</a>.
<a class="reference internal" href="../pyms/Peak/Peak.html#pyms.Peak.List.DPA.Class.PairwiseAlignment" title="pyms.Peak.List.DPA.Class.PairwiseAlignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PairwiseAlignment()</span></code></a>, defined in
<a class="reference internal" href="../pyms/Peak/Peak.html#module-pyms.Peak.List.DPA.Class" title="pyms.Peak.List.DPA.Class"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyms.Peak.List.DPA.Class</span></code></a>
is a class that calculates the similarity between all peaks in
one sample with those of another sample.  This is done for all
possible pairwise alignments (2-alignments). The output of
<a class="reference internal" href="../pyms/Peak/Peak.html#pyms.Peak.List.DPA.Class.PairwiseAlignment" title="pyms.Peak.List.DPA.Class.PairwiseAlignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PairwiseAlignment()</span></code></a>,
(<code class="docutils literal notranslate"><span class="pre">T1</span></code>) is an object which contains the dendrogram tree that
maps the similarity relationship between the input 1-alignments,
and also 1-alignments themselves.</p>
<p>The function <code class="xref py py-meth docutils literal notranslate"><span class="pre">align_with_tree()</span></code>
takes the object T1 and aligns the individual alignment objects according to
the guide tree.  In this example, the individual alignments are
three 1-alignments, and the function
<code class="xref py py-meth docutils literal notranslate"><span class="pre">align_with_tree()</span></code>
first creates a 2-alignment from the two most similar 1-alignments
and then adds the third 1-alignment to this to create a 3-alignment.
The parameter <code class="docutils literal notranslate"><span class="pre">min_peaks=2</span></code> specifies that any peak column of the
data matrix that has less than two peaks in the final alignment
will be dropped.  This is useful to clean up the data matrix of
accidental peaks that are not truly observed over the set of replicates.</p>
<p>Finally, the resulting 3-alignment is saved by writing alignment tables
containing peak retention times (<code class="docutils literal notranslate"><span class="pre">rt1.csv</span></code>) and the corresponding peak areas
(<cite>area1.csv’). These two files are plain ASCII files is CSV format, and are
saved in the directory ``62/output/`</cite>.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">area1.csv</span></code> contains the data matrix where the corresponding peaks are
aligned in the columns and each row corresponds to an experiment. The file
<code class="docutils literal notranslate"><span class="pre">rt1.csv</span></code> is useful for manually inspecting the alignment in some GUI driven
program.</p>
</div>
<div class="section" id="between-state-alignment-of-peak-lists-from-multiple-experiments">
<h2><a class="toc-backref" href="#id10">Between-state alignment of peak lists from multiple experiments</a><a class="headerlink" href="#between-state-alignment-of-peak-lists-from-multiple-experiments" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This example is in <a class="reference external" href="../pyms-test/63/63.html">pyms-test/63</a></p>
</div>
<p>In the previous example the list of peaks were aligned within a single
experiment with multiple replicates (“within-state alignment”).  In practice,
it is of more interest to compare the two experimental states. In a typical
experimental setup there can be multiple replicate experiments on each
experimental state or condition. To analyze the results of such an experiment
statistically, the list of peaks need to be aligned within each
experimental state and also between the states. The result of such an alignment
would be the data matrix of integrated peak areas. The data matrix contains
a row for each sample and the number of columns is determined by
the number of unique peaks (metabolites) detected in all the experiments.</p>
<p>In principle, all experiments could be aligned across conditions and
replicates in the one process. However, a more robust approach is to first align
experiments within each set of replicates (within-state
alignment), and then to align the resulting alignments (between-state
alignment) <a class="footnote-reference brackets" href="#id4" id="id3">1</a>.</p>
<p>This example demonstrates how the peak lists from two cell states are
aligned. The cell state, A, consisting of three experiments aligned in
<code class="docutils literal notranslate"><span class="pre">pyms-test/61a</span></code> (<code class="docutils literal notranslate"><span class="pre">a0806_077</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_078</span></code>, <code class="docutils literal notranslate"><span class="pre">a0806_079</span></code>) and cell state,
B, consisting of three experiments aligned in <code class="docutils literal notranslate"><span class="pre">pyms-test/61b</span></code> (<code class="docutils literal notranslate"><span class="pre">a0806_140</span></code>,
<code class="docutils literal notranslate"><span class="pre">a0806_141</span></code>, and <code class="docutils literal notranslate"><span class="pre">a0806_142</span></code>).</p>
<p>The between group alignment can be performed by the following alignment
commands.</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="c1"># between replicates alignment parameters</span>
<span class="n">Db</span> <span class="o">=</span> <span class="mf">10.0</span> <span class="c1"># rt modulation</span>
<span class="n">Gb</span> <span class="o">=</span> <span class="mf">0.30</span> <span class="c1"># gap penalty</span>

<span class="k">print</span> <span class="s1">&#39;Aligning input {1,2``&#39;</span>
<span class="n">T9</span> <span class="o">=</span> <span class="n">PairwiseAlignment</span><span class="p">([</span><span class="n">A1</span><span class="p">,</span><span class="n">A2</span><span class="p">],</span> <span class="n">Db</span><span class="p">,</span> <span class="n">Gb</span><span class="p">)</span>
<span class="n">A9</span> <span class="o">=</span> <span class="n">align_with_tree</span><span class="p">(</span><span class="n">T9</span><span class="p">)</span>

<span class="n">A9</span><span class="o">.</span><span class="n">write_csv</span><span class="p">(</span><span class="s1">&#39;output/rt.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;output/area.csv&#39;</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
<p>where <code class="docutils literal notranslate"><span class="pre">A1</span></code> and <code class="docutils literal notranslate"><span class="pre">A2</span></code> are the results of the within group alignments
(as above) for group A and B, respectively.</p>
<p>In this example the retention time tolerance for between-state alignment is
greater compared to the retention time tolerance for the within-state alignment
as we expect less fidelity in retention times between them.  The same functions
are used for the within-state and between-state alignment. The result of the
alignment is saved to a file as the area and retention time matrices (described
above).</p>
</div>
<div class="section" id="common-ion-area-quantitation">
<h2><a class="toc-backref" href="#id11">Common Ion Area Quantitation</a><a class="headerlink" href="#common-ion-area-quantitation" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This example is in <a class="reference external" href="../pyms-test/41/41.html">pyms-test/41</a></p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">area.csv</span></code> file produced in the preceding section lists the
total area of each peak in the alignment. The total area is the sum of
the areas of each of the individual ions in the peak. While this approach
produces broadly accurate results, it can result in errors where
neighbouring peaks or unfiltered noise add to the peak in some way.</p>
<p>One alternative to this approach is to pick a single ion which is
common to a particular peak (compound), and to report only the area of this ion
for each occurrence of that peak in the alignment. Using the method
<code class="xref py py-meth docutils literal notranslate"><span class="pre">common_ion()</span></code>
of the PyMS class Alignment, PyMS can select an ion for
each aligned peak which is both abundant and occurs most often
for that peak. We call this the ‘Common Ion Algorithm’ (CIA).</p>
<p>To implement this in PyMS, it is essential that the individual ion
areas have been set (see section ref{sec:individual-ion-areas``).</p>
<div class="section" id="using-the-common-ion-algorithm">
<h3><a class="toc-backref" href="#id12">Using the Common Ion Algorithm</a><a class="headerlink" href="#using-the-common-ion-algorithm" title="Permalink to this headline">¶</a></h3>
<p>When using the CIA for area quantitation, a different method of the
class Alignment is used to write the area matrix;
<code class="xref py py-meth docutils literal notranslate"><span class="pre">write_common_ion_csv()</span></code>.
This requires a list of the common ions for each peak in the alignment.
This list is generated using the Alignment class method
<code class="xref py py-meth docutils literal notranslate"><span class="pre">common_ion()</span></code>.</p>
<p>Continuing from the previous example, the following invokes common ion
filtering on previously created alignment object ‘A9’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">common_ion_list</span> <span class="o">=</span> <span class="n">A9</span><span class="o">.</span><span class="n">common_ion</span><span class="p">()</span>
</pre></div>
</div>
<p>The variable ‘common_ion_list’ is a list of the common ion for each
peak in the alignment. This list is the same length as the
alignment. To write peak areas using common ion quantitation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">A9</span><span class="o">.</span><span class="n">write_common_ion_csv</span><span class="p">(</span><span class="s1">&#39;output/area_common_ion.csv&#39;</span><span class="p">,</span><span class="n">common_ion_list</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="references">
<h2><a class="toc-backref" href="#id13">References</a><a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<dl class="footnote brackets">
<dt class="label" id="id4"><span class="brackets">1</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id2">2</a>,<a href="#id3">3</a>)</span></dt>
<dd><p>Robinson MD, De Souza DP, Keen WW, Saunders EC, McConville MJ, Speed TP, and Likic VA. A dynamic programming approach for the alignment of signal peaks in multiple gas chromatography-mass spectrometry experiments. <cite>BMC Bioinformatics</cite>, 8:419, 2007</p>
</dd>
</dl>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright PyMassSpec Developers

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>